File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();
        int l = 0; 
        int r = 0;
        map<int,int>mp;
        int mx = 0;

        while(r < n) {
            mp[fruits[r]]++;
            if(mp.size() > 2) {
                while(mp.size() > 2) {
                    mp[fruits[l]]--;
                    if(mp[fruits[l]] == 0) {
                        mp.erase(fruits[l]);
                    }
                    l++;

                }
            }
            if(mp.size()<=2) {
                mx = max(mx,r-l+1);
            }
            r++;
        }
        return mx;
        
        
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        from collections import defaultdict
        
        freq = defaultdict(int)
        i = 0
        max_len = 0
        
        for j in range(len(fruits)):
            # Expand the window by adding fruits[j]
            freq[fruits[j]] += 1
            
            # If we have more than 2 types, shrink from the left
            while len(freq) > 2:
                freq[fruits[i]] -= 1
                if freq[fruits[i]] == 0:
                    del freq[fruits[i]]
                i += 1
            
            # Update max window size
            max_len = max(max_len, j - i + 1)
        
        return max_len
